### PR TITLE
Add support for qutrit state preparations 

### LINKED ISSUE #4122: Support qutrit state preparations
### Feature details

PennyLane has growing support for qutrit-based calculations, with the addition of a [`default.qutrit`](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_qutrit.html#module-pennylane.devices.default_qutrit) device and some [qutrit gates](https://docs.pennylane.ai/en/stable/introduction/operations.html#qutrit-operators). One missing element is _state preparation_. We would like to add `QutritBasisState` (analogous to [`BasisState`](https://docs.pennylane.ai/en/stable/code/api/pennylane.BasisState.html)), allowing users to more easily specify the input state to qutrit-based circuits.

### Implementation

`QutritBasisState` should accept:
- `n` (array/tensor): the qutrit basis state, e.g., `[0, 2,  1, 2]` for a 4-qutrit system
- `wires`

The decomposition can be provided using [`TShift`](https://docs.pennylane.ai/en/stable/code/api/pennylane.TShift.html).

This addition should live in a `state_preparation.py` file within [pennylane.ops.qutrit](https://github.com/PennyLaneAI/pennylane/tree/master/pennylane/ops/qutrit) and include documentation, examples, and tests.

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by EmilianoG-byte**
**Context:**
PennyLane has growing support for qutrit-based calculations, with the addition of a [default.qutrit](https://docs.pennylane.ai/en/stable/code/api/pennylane.devices.default_qutrit.html#module-pennylane.devices.default_qutrit) device and some [qutrit gates](https://docs.pennylane.ai/en/stable/introduction/operations.html#qutrit-operators). One missing element is state preparation.

**Description of the Change:**
Add `QutritBasisState` (analogous to [BasisState](https://docs.pennylane.ai/en/stable/code/api/pennylane.BasisState.html)) along with its `decomposition` and `state vector` methods. Together with it, update `default_qutrit` device to accept `apply_basis_state` method.

**Benefits:**
 Allow users to more easily specify the input state to qutrit-based circuits.

**Possible Drawbacks:**
For completeness `QubitDensityMatrix` and `QubitStateVector` should be implemented as well for qutrits.

**Related GitHub Issues:**
Closes #4122

### CODE REVIEW

**Comment by EmilianoG-byte** [comment]
I believe this PR is done. I am happy to receive comments/suggestions about it 😄

**Comment by EmilianoG-byte** [comment]
Should I merge the 2 new commits that have occurred since I opened this PR?

**Comment by trbromley** [comment]
Thanks @EmilianoG-byte! I think you're getting notifications, but in case you are not - there is a discussion in the [main issue thread](https://github.com/PennyLaneAI/pennylane/issues/4122)

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (lillian542) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:117]: This test passes locally for me - was it raising an error for you?

**Review by lillian542 — COMMENTED** [review]
Looking good, great work! 🎉 

Gave it a first pass and left a few comments. 

One thing that is missing is adding the new classes to the documentation; that will also let us confirm that the docs are rendering correctly. Please add the `QutritBasisStatePreparation` template to `docs/introduction/templates.rst`, and the `QutritBasisState` operator to `docs/introduction/operations.rst` in the appropriate sections.

In the tests it seems like most of the parametrizations have been copied over from the qubit version, and don't include preparation in the 2nd excited state (i.e. most of the parametrization of `basis_state` are composed only of 0s and 1s). Could you add additional parametrizations that check this as well?

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:45]: Done!

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:77]: Oh, thank you for the explanation about `is_abstract`! I was actually not sure what the function was doing and I could not access the docs of the qml.math module for some reason. For future references, do you know how to get access to it? Also, if JIT is not compatible with `copmute_decomposition` in general, should we also erase this check from the qubit version: `BasisStatePreparation`?

File: doc/releases/changelog-dev.md
Comment (EmilianoG-byte) [reply doc/releases/changelog-dev.md:223]: thank you for the reminder! :D

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:117]: I actually left it out because I could not check it out locally since I could not find a working installation of Tensorflow for M1 Mac architecture. I also thought `default_qubit_tf/jax` should be implemented for these two tests to make sense. Is that not the case?

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:89]: Done! Just small question: is marking it as "xfail" preferred over "skip"?

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: Well, I am uncommented now, and it is not passing the test. I honestly don't know what can be going wrong since I cannot test it locally and I don't quite understand from the pipeline test where is the error coming from.

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (lillian542) [reply pennylane/templates/state_preparations/basis_qutrit.py]: No worries! Basically, when jitting, you try to make an abstract version of the full function, with placeholders for all the variables that won't be set until the function is called. Then you execute all of the logic once (with the placeholders). After that, you can pass it variables to put instead of the placeholders to get the results. This makes the first execution slow, because it needs to compile everything in a generalized fashion, but makes all subsequent calls much faster, because _almost_ everything is already done. 

So `is_abstract` if a function we use to check if we are currently jitting - we just take one of the objects that we expect to be a represented by a placeholder when jitting, and check if it is a concrete value or a placeholder (abstract). For future reference, you can find the math module documentation under API on the Pennylane.ai documentation page! https://docs.pennylane.ai/en/stable/code/qml_math.html

Wrt whether this is an issue for the qubit implementation, the jitting problem in this case is specific to _this_ implementation of `compute_decomposition`. The qubit version is JIT compatible, because it has a separate clause for generating a tape with parametrized operators if `is_abstract` returns `True`. Basically, operator _parameters_ can be abstract, but an operator's _existence_ can't be. The implementation for qutrits here, which involves changing the number of TShift operators being initialized (a concrete quantity) depending on the target state (an abstract placeholder), can't be made jittable. 

In the future, when more parametrized qutrit operations are added to PennyLane, we can hopefully add JIT compatibility similar to what has been implemented for qubit systems, where a placeholder parameter value is dependent on a placeholder target state value.

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (lillian542) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: You are absolutely right, I ran the qubit test file. The jax and tf interfaces aren't implemented for qutrits, just mark both as xfail!

If you want `tensorflow` on your M1 Mac, I used this install guide for mine: https://caffeinedev.medium.com/how-to-install-tensorflow-on-m1-mac-8e9b91d93706

But I believe you have to be using conda via Miniforge for it to work, I'll leave it up to you whether reconfiguring your setup is worth it just for `tensorflow` 😅

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (lillian542) [inline pennylane/templates/state_preparations/basis_qutrit.py:23]: Please add the `QutritBasisStatePreparation` template to `docs/introduction/templates.rst` so its included in the documentation

File: pennylane/ops/qutrit/state_preparation.py
Comment (lillian542) [inline pennylane/ops/qutrit/state_preparation.py:28]: Please add the `QutritBasisState` operator to `doc/introduction/operations.rst` in the appropriate section, see `BasisState` for reference

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (lillian542) [inline tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:22]: In the tests it seems like most of the parametrizations of `basis_state` are composed of only 0s and 1s. Could you add additional parametrizations that include the 2nd excited state?

**Review by lillian542 — COMMENTED** [review]
@EmilianoG-byte Looking good! A couple more small comments to be addressed and then we can open this up for a second reviewer 🚀

File: pennylane/ops/qutrit/state_preparation.py
Comment (EmilianoG-byte) [reply pennylane/ops/qutrit/state_preparation.py:28]: Done!

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py]: I see, thank you so much for taking the time to explain this! Maybe I will help contribute to the future JIT comptability :)

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: okay! thank you!

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: Done! And thank you for the TF link, this is gonna be super useful! If you have a similar one for JAX that works with Pennylane, that would also be great :) (between numpy, jax, and jaxlib incompatibilities, I found it impossible to get them all working at the same time)

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py:22]: done!

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: How is the png image included in `:figure:` generated?

Comment (lillian542) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: That's not a super set thing. But you can just run this to get an image representing the template, and then save it as a png and add it to the folder:

```
import pennylane as qml

dev = qml.device("default.qutrit", wires=3)

@qml.qnode(dev)
def circuit(basis_state):
    qml.QutritBasisStatePreparation(basis_state, wires=range(3))
    return qml.state()

basis_state = [1, 0, 2]
qml.draw_mpl(circuit)(basis_state)
```

You're also welcome to create a more illustrative circuit using `TShift` operators instead of `QutritBasisStatePreparation` and draw that instead if you prefer, akin to the circuit shown for the `BasisState` template :)

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (lillian542) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: Once I had the conda miniforge stuff set up I just pip-installed those ones, but there can be some issues with versions for sure. I currently have:

tensorflow-macos: 2.12.0
jax and jaxlib: 0.4.10
numpy: 1.23.5

and they seem to all be playing nicely together. I actually changed my numpy version a few days ago because it wasn't getting along with tensorflow 🥲

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [inline pennylane/templates/state_preparations/basis_qutrit.py:48]: I have changed this so it is displayed as "code" rather than just a comment (to have it similar to the examples in other templates). I also changed `BasisStatePreparation` accordingly. I hope this is okay :)

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: I added this graph:

image

but I think one of these two look "friendlier" 😆 

image

image

But I guess that would imply changing the other ones to keep similar format, no?

File: tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py
Comment (EmilianoG-byte) [reply tests/templates/test_state_preparations/test_qutrit_basis_state_prep.py]: It can really become super messy to try to make them work together 😪.
I will keep the link and versions you sent me somewhere safe for when I next need it 😆. Thank you! :D

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: So something is off because the documentation check is failing but the documentation in readthedocs looks just fine.

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (lillian542) [reply pennylane/templates/state_preparations/basis_qutrit.py:48]: Nice!

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (lillian542) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: I agree, the new plotting format is friendlier 😅 But keeping the format consistent makes sense. 

Sphinx can be a little finicky for the docs build, this is the error it seems to be reporting (you just have to scroll up endlessly until you eventually find its complaint, unfortunately it doesn't give any details in the final `build failed` message):

image

File: pennylane/templates/state_preparations/basis.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis.py:49]: Ohhhh, I seeee! I totally did not see that coming. Thank you for pointing out!

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: Omg, thank you! I searched "error" and could not find the source of the problem. And even the warning on your screenshot showed "line 27" which is not even where the error was 😕  Thank you for pointing it out!! 

To build the docs locally for testing, I understand from this https://docs.pennylane.ai/en/stable/development/guide/documentation.html that I should have the interfaces requirements installed, right?

Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:23]: I see, once again: thank you 😄 
I believe this was the last fix from this batch of comments then!

File: pennylane/devices/default_qutrit.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit.py:135]:
```suggestion
                    f"Operation {operation.name} cannot be used after other operations have already been applied "
```

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (mudit2812) [inline pennylane/templates/state_preparations/basis_qutrit.py:86]: If this will not be addressed in this PR, could you please open an issue for this so that the core team can track it and hopefully make changes to make this work?

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (mudit2812) [inline pennylane/templates/state_preparations/basis_qutrit.py:117]: Nice! As a side note, using `qml.adjoint(qml.TShift(wire))` would also be a viable option for making the `|0>` to `|2>` transition instead of two separate `TShift` operations. No need to add this change, but just a random thought that popped up.

File: pennylane/devices/default_qutrit.py
Comment (mudit2812) [inline pennylane/devices/default_qutrit.py:167]: I think this check would be more appropriate in the `__init__` method of `QutritBasisState`. The error shouldn't rely on the device, as it is due to an ill-defined operator as opposed to an ill-defined device.

As an example, if someone created a `QutritBasisState` instance outside a `QNode` or quantum tape with the parameter `[0, 1, 2, 3]`, an error wouldn't get thrown, but `QutritBasisState.state_vector` would likely not work.

File: pennylane/ops/qutrit/state_preparation.py
Comment (mudit2812) [inline pennylane/ops/qutrit/state_preparation.py:95]: Following up on my previous suggestion to do this check in the `__init__` method of this class, I would remove this check from here if you do make the suggested change as this check would then be redundant.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Thanks for your contribution @EmilianoG-byte ! This PR looks really good. I left some minor comments and a couple of suggestions that I think would be helpful to make this class work better outside of queuing contexts (qnodes, quantum tapes, annotated queues, etc.). After addressing those changes, I think this should be ready to merge!

File: pennylane/devices/default_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/devices/default_qutrit.py:167]: You are right, an error is not thrown until the `state_vector` method is called. Are you suggesting creating an ` __init__ ` that supers the one inherited from the `Operation` class? Then I should also change everything accordingly for the `Qubit BasisState` preparation and `default.qubit` as well right?

File: pennylane/templates/state_preparations/basis_qutrit.py
Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:86]: I have created the issue with the information I have: #4223

Comment (EmilianoG-byte) [reply pennylane/templates/state_preparations/basis_qutrit.py:86]: I see actually this is issue with not having `state` as a hyper parameter is somewhat related to your comment regarding the assertion checking, as these operators don't have an __init__ on their own but only inherited from Operations. 

I found this on the `adding_operators.rst`:

```
*Input checks.* Checking the inputs of the operation introduces an overhead and clashes with tools like
  just-in-time compilation. Find a balance of adding meaningful sanity checks (such as for the shape of tensors),
  but keeping them to a minimum.
```

Could this explain why of the absence of these checks?

**Review by mudit2812 — APPROVED** [review]
Thanks for addressing my comments! I think that in it's current state, the PR is merge ready. Any changes pertaining to where we check whether the basis state is valid can be done in a subsequent PR. Great work!

**Review by lillian542 — APPROVED** [review]
Great work @EmilianoG-byte, thanks for contributing! 🎉 🚀

**Comment by EmilianoG-byte** [comment]
Thank you @lillian542 and @mudit2812! I really appreciate the comments and sources I got through the review 😄 Really glad I can contribute, and hopefully I will contributing in the future :)

**Comment by EmilianoG-byte** [comment]
Also I just checked and my user does not appear as one of the contributors on GitHub, is that normal (this is of course not important, but would have been nice to see me up there 😄)

**Comment by trbromley** [comment]
Hi @EmilianoG-byte. Congrats on getting this merged and thanks for contributing to PennyLane! :tada:

> Also I just checked and my user does not appear as one of the contributors on GitHub

Do you mean on the [contributor page](https://github.com/PennyLaneAI/pennylane/graphs/contributors)? That page is automated, but [here](https://docs.github.com/en/repositories/viewing-activity-and-data-for-your-repository/viewing-a-projects-contributors#troubleshooting-contributors) are some of the reasons you might not be showing up. In this case, I think it's that we have lots of contributors with one commit such that new contributors with one commit do not make it into the top 100 :disappointed:.

Also, finally, as part of our 0.31 release we're planning to promote UnitaryHack contributions to PennyLane on social media. This is completely optional, but would you like to be mentioned in our social media posts? Specifically, we'd need either your Twitter profile, your LinkedIn profile, or your name.

**Comment by EmilianoG-byte** [comment]
Thank you for your answer @trbromley!

I see 😮. I will have to wait for my next contributions until I show up there then 😄 

And regarding the post, that would be really nice actually. This is my [LinkedIn profile](https://www.linkedin.com/in/cristian-emiliano-godinez-ramirez/) (everyone in this thread please feel free to connect :D)

I will make sure to repost ;) Thank you!

**Pull request merged**
